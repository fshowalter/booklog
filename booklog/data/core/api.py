from __future__ import annotations

from booklog.data.core import json_authors, json_works, orm, queries

Work = orm.Work

Author = orm.Author

AuthorWithWorks = orm.AuthorWithWorks

WorkAuthor = orm.WorkAuthor

WORK_KINDS = json_works.KINDS

search_authors = queries.search_authors

search_works = queries.search_works

create_work = orm.create_work

create_author = orm.create_author

all_works = queries.all_works

all_authors = queries.all_authors

all_json_works = json_works.deserialize_all

all_json_authors = json_authors.deserialize_all
