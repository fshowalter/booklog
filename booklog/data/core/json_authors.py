from __future__ import annotations

import json
import os
from glob import glob
from typing import Any, TypedDict

from slugify import slugify

from booklog.data.utils import path_tools
from booklog.logger import logger

FOLDER_NAME = "authors"


JsonAuthor = TypedDict(
    "JsonAuthor",
    {
        "slug": str,
        "name": str,
        "sortName": str,
    },
)


def generate_sort_name(name: str) -> str:
    split_name = name.split()
    last_name = split_name[-1]
    other_names = split_name[:-1]

    return "{0}, {1}".format(last_name, " ".join(other_names))


def author_for_slug(slug: str, all_json_authors: list[JsonAuthor]) -> JsonAuthor:
    return next(author for author in all_json_authors if author["slug"] == slug)


def create(
    name: str,
) -> JsonAuthor:
    slug = slugify(name)

    json_author = JsonAuthor(name=name, sortName=generate_sort_name(name), slug=slug)

    serialize(json_author=json_author)

    return json_author


def deserialize_json_author(json_author: dict[str, Any]) -> JsonAuthor:
    return JsonAuthor(
        name=json_author["name"],
        sortName=json_author["sortName"],
        slug=json_author["slug"],
    )


def deserialize_all() -> list[JsonAuthor]:
    authors: list[JsonAuthor] = []

    for file_path in glob(os.path.join(FOLDER_NAME, "*.json")):
        with open(file_path, "r") as json_file:
            authors.append(deserialize_json_author(json.load(json_file)))

    return authors


def serialize(json_author: JsonAuthor) -> None:
    file_path = os.path.join(FOLDER_NAME, "{0}.json".format(json_author["slug"]))
    path_tools.ensure_file_path(file_path)

    with open(file_path, "w") as output_file:
        output_file.write(json.dumps(json_author, default=str, indent=2))

    logger.log(
        "Wrote {}.",
        file_path,
    )
