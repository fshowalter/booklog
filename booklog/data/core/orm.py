from __future__ import annotations

from dataclasses import dataclass
from typing import Optional

from booklog.data.core import json_authors, json_works


@dataclass
class Author(object):
    name: str
    sort_name: str
    slug: str


@dataclass
class AuthorWithWorks(Author):
    works: list[Work]


@dataclass
class WorkAuthor(Author):
    notes: Optional[str]


@dataclass
class Work(object):
    title: str
    subtitle: Optional[str]
    year: str
    sort_title: str
    authors: list[WorkAuthor]
    slug: str
    kind: str
    included_work_slugs: list[str]
    included_in_work_slugs: list[str]


def create_work(  # noqa: WPS211
    title: str,
    subtitle: Optional[str],
    year: str,
    work_authors: list[json_works.CreateWorkAuthor],
    kind: str,
    included_work_slugs: Optional[list[str]] = None,
) -> Work:
    return hydrate_json_work(
        json_works.create(
            title=title,
            subtitle=subtitle,
            year=year,
            work_authors=work_authors,
            kind=kind,
            included_work_slugs=included_work_slugs,
        )
    )


def create_author(
    name: str,
) -> Author:
    return hydrate_json_author(json_authors.create(name=name))


def hydrate_json_work_authors(
    json_work_authors: list[json_works.JsonWorkAuthor],
) -> list[WorkAuthor]:
    work_authors = []

    for work_author in json_work_authors:
        json_author = json_authors.author_for_slug(slug=work_author["slug"])
        work_authors.append(
            WorkAuthor(
                name=json_author["name"],
                slug=json_author["slug"],
                sort_name=json_author["sortName"],
                notes=work_author["notes"],
            )
        )

    return work_authors


def hydrate_json_work(json_work: json_works.JsonWork) -> Work:
    return Work(
        title=json_work["title"],
        subtitle=json_work["subtitle"],
        sort_title=json_work["sortTitle"],
        year=json_work["year"],
        authors=hydrate_json_work_authors(json_work["authors"]),
        slug=json_work["slug"],
        kind=json_work["kind"],
        included_work_slugs=json_work["includedWorks"],
        included_in_work_slugs=[
            collection_work["slug"]
            for collection_work in json_works.works_including_work_slug(
                json_work["slug"]
            )
        ],
    )


def hydrate_json_author(json_author: json_authors.JsonAuthor) -> Author:
    return Author(
        name=json_author["name"],
        slug=json_author["slug"],
        sort_name=json_author["sortName"],
    )


def hydrate_json_author_with_works(
    json_author: json_authors.JsonAuthor,
) -> AuthorWithWorks:
    return AuthorWithWorks(
        name=json_author["name"],
        slug=json_author["slug"],
        sort_name=json_author["sortName"],
        works=[
            hydrate_json_work(json_work)
            for json_work in json_works.works_for_author_slug(slug=json_author["slug"])
        ],
    )
