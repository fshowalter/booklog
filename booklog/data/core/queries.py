from __future__ import annotations

from booklog.data.core import json_authors, json_works, orm


def all_works() -> list[orm.Work]:
    return [
        orm.hydrate_json_work(json_work=json_work)
        for json_work in json_works.deserialize_all()
    ]


def all_authors() -> list[orm.AuthorWithWorks]:
    return [
        orm.hydrate_json_author_with_works(json_author=json_author)
        for json_author in json_authors.deserialize_all()
    ]


def search_authors(query: str) -> list[orm.AuthorWithWorks]:
    filtered_authors = filter(
        lambda json_author: query.lower() in json_author["name"].lower(),
        json_authors.deserialize_all(),
    )

    return [
        orm.hydrate_json_author_with_works(json_author=json_author)
        for json_author in filtered_authors
    ]


def search_works(query: str) -> list[orm.Work]:
    filtered_works = filter(
        lambda json_work: query.lower()
        in "{0}: {1}".format(json_work["title"], json_work["subtitle"]).lower(),
        json_works.deserialize_all(),
    )

    return [orm.hydrate_json_work(json_work=json_work) for json_work in filtered_works]


# def author_names_for_work(work: Work) -> list[str]:
#     author_slugs = list(map(lambda author: author.slug, work.authors))

#     filtered_authors = filter(lambda author: author.slug in author_slugs, all_authors())

#     return list(map(lambda author: author.name, filtered_authors))


# def author_for_slug(author_slug: str) -> orm.Author:
#     return next(author for author in all_authors() if author.slug == author_slug)


# def works_for_author(author_slug: str) -> list[Work]:
#     return list(
#         filter(
#             lambda work: author_slug in set(author.slug for author in work.authors),
#             all_works(),
#         )
#     )
