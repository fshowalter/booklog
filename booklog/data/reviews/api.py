from __future__ import annotations

from booklog.data.reviews import orm, queries

Review = orm.Review

all_reviews = queries.all_reviews

create_or_update = orm.create_or_update
