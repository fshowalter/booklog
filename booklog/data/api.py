from __future__ import annotations

from booklog.data.core import api as core_api
from booklog.data.exports import api as exports_api
from booklog.data.readings import api as readings_api
from booklog.data.reviews import api as reviews_api

AuthorWithWorks = core_api.AuthorWithWorks

Work = core_api.Work

Author = core_api.Author

WorkAuthor = core_api.WorkAuthor

Review = reviews_api.Review

Reading = readings_api.Reading

TimelineEntry = readings_api.TimelineEntry

create_author = core_api.create_author

search_authors = core_api.search_authors

search_works = core_api.search_works

create_work = core_api.create_work

all_editions = readings_api.all_editions

WORK_KINDS = core_api.WORK_KINDS

export_data = exports_api.export_data


def create_reading(
    work: core_api.Work,
    edition: str,
    timeline: list[TimelineEntry],
    grade: str,
) -> None:
    readings_api.create_reading(
        work=work,
        edition=edition,
        timeline=timeline,
    )

    reviews_api.create_or_update(
        work=work,
        date=timeline[-1].date,
        grade=grade,
    )
