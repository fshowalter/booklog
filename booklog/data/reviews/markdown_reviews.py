from __future__ import annotations

import os
import re
from dataclasses import dataclass
from glob import glob
from typing import Any, Optional, TypedDict

import yaml

from booklog.data.utils import path_tools
from booklog.logger import logger

FOLDER_NAME = "reviews"

FM_REGEX = re.compile(r"^-{3,}\s*$", re.MULTILINE)


def represent_none(self: Any, _: Any) -> Any:
    return self.represent_scalar("tag:yaml.org,2002:null", "")


yaml.add_representer(type(None), represent_none)


@dataclass(kw_only=True)
class MarkdownReview(object):
    yaml: ReviewYaml
    review_content: Optional[str] = None


class ReviewYaml(TypedDict):
    work_slug: str
    grade: str
    date: str


def review_for_work_slug(
    work_slug: str, all_markdown_reviews: list[MarkdownReview]
) -> MarkdownReview:
    return next(
        markdown_review
        for markdown_review in all_markdown_reviews
        if markdown_review.yaml["work_slug"] == work_slug
    )


def create(
    work_slug: str,
    grade: str,
    date: str,
) -> MarkdownReview:
    markdown_review = MarkdownReview(
        yaml=ReviewYaml(
            work_slug=work_slug,
            grade=grade,
            date=date,
        )
    )

    serialize(markdown_review)

    return markdown_review


def deserialize(file_path: str) -> MarkdownReview:
    with open(file_path, "r") as review_file:
        _, frontmatter, review_content = FM_REGEX.split(review_file.read(), 2)

    review_yaml = yaml.safe_load(frontmatter)

    return MarkdownReview(
        yaml=ReviewYaml(
            work_slug=review_yaml["work_slug"],
            grade=review_yaml["grade"],
            date=review_yaml["date"],
        ),
        review_content=review_content,
    )


def deserialize_all() -> list[MarkdownReview]:
    reviews: list[MarkdownReview] = []
    for review_file_path in glob(os.path.join(FOLDER_NAME, "*.md")):
        reviews.append(deserialize(review_file_path))

    return reviews


def generate_file_path(markdown_review: MarkdownReview) -> str:
    file_path = os.path.join(
        FOLDER_NAME, "{0}.md".format(markdown_review.yaml["work_slug"])
    )

    path_tools.ensure_file_path(file_path)

    return file_path


def serialize(markdown_review: MarkdownReview) -> str:
    file_path = generate_file_path(markdown_review)

    stripped_content = str(markdown_review.review_content or "").strip()

    with open(file_path, "w") as output_file:
        output_file.write("---\n")
        yaml.dump(
            markdown_review.yaml,
            encoding="utf-8",
            allow_unicode=True,
            default_flow_style=False,
            sort_keys=False,
            stream=output_file,
        )
        output_file.write("---\n\n")
        output_file.write(stripped_content)

    logger.log("Wrote {}", file_path)

    return file_path
