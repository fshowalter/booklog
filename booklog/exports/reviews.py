import datetime
from typing import Optional, TypedDict

from booklog.bookdata.api import Author, Work, WorkAuthor
from booklog.readings.reading import Reading
from booklog.reviews.review import Review
from booklog.utils import export_tools, list_tools
from booklog.utils.logging import logger

TimelineEntry = TypedDict("TimelineEntry", {"date": datetime.date, "progress": str})

JsonWorkAuthor = TypedDict(
    "JsonWorkAuthor",
    {"name": str, "slug": str, "notes": Optional[str]},
)

JsonTimelineEntry = TypedDict("JsonTimelineEntry", {})


JsonReviewedWork = TypedDict(
    "JsonReviewedWork",
    {
        "sequence": int,
        "workSlug": str,
        "includedInWorkSlugs": list[str],
        "title": str,
        "yearPublished": str,
        "edition": str,
        "authors": list[JsonWorkAuthor],
        "grade": str,
        "kind": str,
        "date": datetime.date,
    },
)


def build_json_work_authors(
    work_authors: list[WorkAuthor], authors: list[Author]
) -> list[JsonWorkAuthor]:
    json_work_authors = []

    for work_author in work_authors:
        author_name = next(
            author.name for author in authors if author.slug in work_author.slug
        )

        json_work_authors.append(
            JsonWorkAuthor(
                slug=work_author.slug, notes=work_author.notes, name=author_name
            )
        )

    return json_work_authors


def build_json_reviewed_work(
    readings: list[Reading],
    authors: list[Author],
    work: Work,
    included_in_work_slugs: list[str],
    review: Review,
) -> JsonReviewedWork:
    most_recent_reading = sorted(
        readings, key=lambda reading: reading.sequence, reverse=True
    )[0]

    return JsonReviewedWork(
        sequence=most_recent_reading.sequence,
        workSlug=review.work_slug,
        title=work.title,
        yearPublished=work.year,
        edition=most_recent_reading.edition,
        grade=review.grade,
        kind=work.kind,
        date=sorted(
            most_recent_reading.timeline, key=lambda entry: entry.date, reverse=True
        )[0].date,
        authors=build_json_work_authors(work_authors=work.authors, authors=authors),
        includedInWorkSlugs=included_in_work_slugs,
    )


def export(
    readings: list[Reading],
    authors: list[Author],
    works: list[Work],
    reviews: list[Review],
) -> None:
    logger.log("==== Begin exporting {}...", "reviews")

    json_reviewed_works = []

    works_by_slug = list_tools.list_to_dict_by_key(works, lambda work: work.slug)
    readings_by_work_slug = list_tools.group_list_by_key(
        readings, lambda reading: reading.work_slug
    )

    for review in reviews:
        if not readings_by_work_slug[review.work_slug]:
            continue

        json_reviewed_works.append(
            build_json_reviewed_work(
                readings=readings_by_work_slug[review.work_slug],
                authors=authors,
                work=works_by_slug[review.work_slug],
                review=review,
                included_in_work_slugs=[
                    work.slug
                    for work in works
                    if review.work_slug in work.included_works
                ],
            )
        )

    export_tools.serialize_dicts(
        sorted(json_reviewed_works, key=lambda work: work["sequence"], reverse=True),
        "updates",
    )
