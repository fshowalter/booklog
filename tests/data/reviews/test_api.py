import os
from datetime import date
from pathlib import Path

import pytest
from syrupy.assertion import SnapshotAssertion

from booklog.data.core import api as core_api
from booklog.data.reviews import api as reviews_api


@pytest.fixture
def created_author() -> core_api.Author:
    return core_api.create_author("Richard Laymon")


@pytest.fixture
def created_work(created_author: core_api.Author) -> core_api.Work:
    return core_api.create_work(
        title="The Cellar",
        subtitle=None,
        year="1980",
        kind="Novel",
        included_work_slugs=[],
        work_authors=[
            core_api.WorkAuthor(
                name=created_author.name,
                sort_name=created_author.sort_name,
                slug=created_author.slug,
                notes=None,
            )
        ],
    )


def test_can_create_new_review(
    tmp_path: Path, created_work: core_api.Work, snapshot: SnapshotAssertion
) -> None:
    reviews_api.create_or_update(
        work=created_work,
        grade="A+",
        date=date(2016, 3, 10),
    )

    with open(
        os.path.join(tmp_path / "reviews", "the-cellar-by-richard-laymon.md"), "r"
    ) as output_file:
        file_content = output_file.read()

    assert file_content == snapshot


def test_can_update_existing_review(
    tmp_path: Path, created_work: core_api.Work, snapshot: SnapshotAssertion
) -> None:
    existing_review = "---\nwork_slug: the-cellar-by-richard-laymon\ngrade: A+\ndate: 2016-03-10\n---\n\nSome review content we want to preserve between updates."  # noqa: 501

    with open(
        os.path.join(tmp_path / "reviews", "the-cellar-by-richard-laymon.md"),
        "w",
    ) as first_output_file:
        first_output_file.write(existing_review)

    reviews_api.create_or_update(
        work=created_work,
        grade="C+",
        date=date(2017, 3, 12),
    )

    with open(
        os.path.join(tmp_path / "reviews", "the-cellar-by-richard-laymon.md"), "r"
    ) as output_file:
        file_content = output_file.read()

    assert file_content == snapshot
