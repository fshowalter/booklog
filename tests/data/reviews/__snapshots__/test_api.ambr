# serializer version: 1
# name: test_can_create_new_review
  '''
  ---
  work_slug: the-cellar-by-richard-laymon
  grade: A+
  date: '2016-03-10'
  ---
  
  
  '''
# ---
# name: test_can_update_existing_review
  '''
  ---
  work_slug: the-cellar-by-richard-laymon
  grade: C+
  date: '2017-03-12'
  ---
  
  Some review content we want to preserve between updates.
  '''
# ---
